cmake: require python 3.6 or later

mbolivar opened the pull request:
Zephyr currently requires Python 3.4 or later. The core Python team
declared version 3.4 hit End of Life (EOL) in March, so there's no
reason to continue to support it if that's causing a burden, which it
is.

This commit allows Zephyr's Python scripts to depend on features
present in version 3.6 or later.

This does skip support for a currently active version of Python:

- Python 3.5 is actively supported by the core Python devs until 09/2020
- Zephyr's 2.2 release, the first which could include this change, is
  tentatively scheduled for 02/2020.

However, almost all supported platforms are either unaffected, or
their users can upgrade easily:

- Windows users who need to can upgrade Python with:
  choco upgrade python

- macOS users who need to can upgrade Python with:
  brew upgrade python3

- Red Hat Enterprise Linux users who need to upgrade can use
  Software Collections (SCLs), e.g. as described here:
  https://developers.redhat.com/blog/2018/08/13/install-python3-rhel/

- CentOS Linux users also have access to SCLs, as described here:
  https://wiki.centos.org/AdditionalResources/Repositories/SCL

- Ubuntu's current long-term support (LTS) release (Bionic Beaver,
  version 18.04) ships with Python 3.6. It and all later versions of
  Ubuntu won't be affected by this change.

- Debian's current stable release (Buster, version 10) ships Python 3.7
  and likewise won't be affected.

The impact of this change is therefore biggest for older versions of
Linux. In particular, these are impacted:

- Older Ubuntu LTS releases.

  - Ubuntu 16.04 ships Python 3.5; it is still supported by Canonical.
  - Ubuntu 14.04 ships Python 3.4, which is EOL. This Ubuntu version
    is also no longer getting standard support from Canonical. Paying
    customers are receiving security updates only.

  https://wiki.ubuntu.com/Releases

- Older Debian versions.

  - Debian 9 (stretch) ships Python 3.5 and is still a supported
    Debian version.
  - Debian 8 (jessie) ships Python 3.4, which is EOL. This Debian
    version is no longer receiving mainline maintenance by the Debian
    project. LTS updates are provided by interested community
    volunteers only.

  https://wiki.debian.org/LTS

Affected Linux users will no longer have a system Python 3 which works
"out of the box" with Zephyr after this change. Some ideas for these
users are:

- Use Zephyr v2.1 or v1.14 LTS, which are maintained and still
  support Python 3.4
- Compile Python 3.6 or later from source and use it within a venv:
  https://docs.python.org/3/library/venv.html
- Use something like https://github.com/pyenv/pyenv

Python 3.6 has compelling new features which make writing Zephyr's
scripts easier, and which it would be good to be able to rely upon.
This motivates moving from Python 3.4 to 3.6 instead of 3.5.

My personal killer 3.6 features motivating skipping 3.5 (YMMV):

- Windows console and file system encodings are UTF-8 (PEPs 528 and
  529): Zephyr's scripts, and many utilities related to git, broadly
  assume strings are UTF-8, so this is very helpful
- os.PathLike and the file system path protocol (PEP 519) allow
  intermixing "smart" paths in pathlib with existing os.path based
  code
- f-strings (PEP 0498) are a wonderful and efficient string
  interpolation mechanism
- CPython dictionaries are insertion ordered as an implementation
  detail starting with 3.6, which in practice helps with
  reproducibility (and *all* Python implementations have insertion
  ordered dicts starting with 3.7)

galak committed:
ci: use v0.10.1 docker image

Bump to v0.10.1 Docker Image to get access to SDK 0.11.0-alpha8,
Ubuntu 18.04, gcc-9, clang-9, and python 3.6.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

mbolivar-nordic committed:
cmake: require python 3.6 or later

Zephyr currently requires Python 3.4 or later. The core Python team
declared version 3.4 hit End of Life (EOL) in March, so there's no
reason to continue to support it if that's causing a burden, which it
is.

This commit allows Zephyr's Python scripts to depend on features
present in version 3.6 or later.

This does skip support for a currently active version of Python:

- Python 3.5 is actively supported by the core Python devs until 09/2020
- Zephyr's 2.2 release, the first which could include this change, is
  tentatively scheduled for 02/2020.

However, almost all supported platforms are either unaffected, or
their users can upgrade easily:

- Windows users who need to can upgrade Python with:
  choco upgrade python

- macOS users who need to can upgrade Python with:
  brew upgrade python3

- Red Hat Enterprise Linux users who need to upgrade can use
  Software Collections (SCLs), e.g. as described here:
  https://developers.redhat.com/blog/2018/08/13/install-python3-rhel/

- CentOS Linux users also have access to SCLs, as described here:
  https://wiki.centos.org/AdditionalResources/Repositories/SCL

- Ubuntu's current long-term support (LTS) release (Bionic Beaver,
  version 18.04) ships with Python 3.6. It and all later versions of
  Ubuntu won't be affected by this change.

- Debian's current stable release (Buster, version 10) ships Python 3.7
  and likewise won't be affected.

The impact of this change is therefore biggest for older versions of
Linux. In particular, these are impacted:

- Older Ubuntu LTS releases.

  - Ubuntu 16.04 ships Python 3.5; it is still supported by Canonical.
  - Ubuntu 14.04 ships Python 3.4, which is EOL. This Ubuntu version
    is also no longer getting standard support from Canonical. Paying
    customers are receiving security updates only.

  https://wiki.ubuntu.com/Releases

- Older Debian versions.

  - Debian 9 (stretch) ships Python 3.5 and is still a supported
    Debian version.
  - Debian 8 (jessie) ships Python 3.4, which is EOL. This Debian
    version is no longer receiving mainline maintenance by the Debian
    project. LTS updates are provided by interested community
    volunteers only.

  https://wiki.debian.org/LTS

Affected Linux users will no longer have a system Python 3 which works
"out of the box" with Zephyr after this change. Some ideas for these
users are:

- Use Zephyr v2.1 or v1.14 LTS, which are maintained and still
  support Python 3.4
- Compile Python 3.6 or later from source and use it within a venv:
  https://docs.python.org/3/library/venv.html
- Use something like https://github.com/pyenv/pyenv

Python 3.6 has compelling new features which make writing Zephyr's
scripts easier, and which it would be good to be able to rely upon.
This motivates moving from Python 3.4 to 3.6 instead of 3.5.

My personal killer 3.6 features motivating skipping 3.5 (YMMV):

- Windows console and file system encodings are UTF-8 (PEPs 528 and
  529): Zephyr's scripts, and many utilities related to git, broadly
  assume strings are UTF-8, so this is very helpful
- os.PathLike and the file system path protocol (PEP 519) allow
  intermixing "smart" paths in pathlib with existing os.path based
  code
- f-strings (PEP 0498) are a wonderful and efficient string
  interpolation mechanism
- CPython dictionaries are insertion ordered as an implementation
  detail starting with 3.6, which in practice helps with
  reproducibility (and *all* Python implementations have insertion
  ordered dicts starting with 3.7)

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

galak commented:
You need to update the docker image for CI as we have python 3.5 installed there.

carlescufi reviewed (approved):
Agreed, no reason to support anything below 3.6 anymore.

nashif commented:
> You need to update the docker image for CI as we have python 3.5 installed there.

that is going to be fun, I think it is time to update the docker image OS as well, it has been a while and it is running an ancient ubuntu.

nashif commented:
Why the TSC label? Is there any controversy around this already?

mbolivar commented:
> Why the TSC label? Is there any controversy around this already?

Just being cautious here. I wanted to make sure project members have a heads-up by discussing it briefly (1-2 min) at the next TSC meeting. I can drop this label if people don't think that's necessary.

mbolivar commented:
> You need to update the docker image for CI as we have python 3.5 installed there.

https://github.com/zephyrproject-rtos/ci-dockerfiles/pull/63

nashif commented:
> that is going to be fun, I think it is time to update the docker image OS as well, it has been a while and it is running an ancient ubuntu.

ah, @galak already updated that a while back, cool.

ulfalizer commented:
Don't care all that much about the path stuff myself (`pathlib` is available since 3.4, and mixing with `os` isn't super painful). f-strings are nice though: Can write `f"foo = {foo.bar}, bar = {baz}"` instead of `"foo ={}, bar = {}".format(foo.bar, baz)"`.

I might've waited a bit longer (Python 3.5 EOLs 2020-09-13), but up to you.

aescolar commented:
Thanks @mbolivar. Giving reasonable warning is always nice.
One comment: note that corporate environments tend to use oldish distros, with RedHat being common. I just quickly checked and RedHat 7, which is probably the most common RedHat today, seems to provide 3.6 thru an extra RedHat repo.

SebastianBoe commented:
The 18.04 ubuntu release is quite recent (3 months old), but the python release is quite old, so this seems reasonable to me.

This will most likely not affect Windows users (due to 3.6 being released so long ago).

Ubuntu release schedule: https://wiki.ubuntu.com/Releases
Python releases: https://en.wikibooks.org/wiki/Python_Programming/Version_history

aescolar reviewed (commented):
No complaints from me, as it is not too overly new.

mbolivar commented:
> The 18.04 ubuntu release is quite recent (3 months old), but the python release is quite old, so this seems reasonable to me.
> [...]
> https://wiki.ubuntu.com/Releases

Thanks for the review!

Just a quick note that Ubuntu's release numbers are calendar-based, so 18.04 was released in 2018 (in April, month 4). So this latest LTS has actually been out for over a year now. You might be looking at 18.04.3, the latest point release, instead of the first 18.04 line in that link.

SebastianBoe commented:
Aha, that's much better, thanks :)

galak commented:
Going to put out an -rc of the docker image for testing.  We use this for both zephyr CI and SDK builds so want to make sure things work for both before we update.

galak commented:
updated #20625 for testing with 0.10.0-rc1 of the ci image.

mbolivar-nordic committed:
doc: update GSG and Linux guides for move to Python 3.6

Add information about the minimum Python version to the advanced Linux
documentation.

Drop 16.04 from the GSG since its system Python 3 is no longer covered
by these instructions.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar commented:
- Rebased on 2.1
- Updated Linux documentation
- Updated PR description to more fully describe impact and upgrade mechanisms

mgti commented:
Looks good - +1

dbkinder reviewed (approved):
Thanks for updating the GSG...

mbolivar commented:
@galak That seems like a long enough comment period without any controversy; I think this is ready to go in.

mbolivar commented:
> @galak That seems like a long enough comment period without any controversy; I think this is ready to go in.

I think he's on vacation -- can anybody else merge this?

mbolivar commented on .shippable.yml:
```diff
@@ -20,7 +20,7 @@ build:
     pre_ci_boot:
         image_name: zephyrprojectrtos/ci
-        image_tag: v0.9.1
+        image_tag: v0.10.0-rc1
```
I rebased on top of that PR instead. Thanks.

andrewboie commented:
> I think he's on vacation -- can anybody else merge this?

Sure, once CI is passing I'll push it.
